hazelcast/src/main/java/com/hazelcast/jet/impl/execution/init/ExecutionPlanBuilder.java

```diff
@@ -69,25 +72,49 @@ private ExecutionPlanBuilder() {
 
     @SuppressWarnings({"checkstyle:ParameterNumber", "rawtypes"})
     public static CompletableFuture<Map<MemberInfo, ExecutionPlan>> createExecutionPlans(
-            NodeEngineImpl nodeEngine, List<MemberInfo> memberInfos, DAG dag, long jobId, long executionId,
-            JobConfig jobConfig, long lastSnapshotId, boolean isLightJob, Subject subject
+            NodeEngineImpl nodeEngine,
+            List<MemberInfo> memberInfos,
+            DAG dag,
+            long jobId,
+            long executionId,
+            JobConfig jobConfig,
+            long lastSnapshotId,
+            boolean isLightJob,
+            Subject subject
     ) {
         final VerticesIdAndOrder verticesIdAndOrder = VerticesIdAndOrder.assignVertexIds(dag);
         final int defaultParallelism = nodeEngine.getConfig().getJetConfig().getCooperativeThreadCount();
-        final Map<MemberInfo, int[]> partitionsByMember = getPartitionAssignment(nodeEngine, memberInfos);
-        final Map<Address, int[]> partitionsByAddress =
+        Map<MemberInfo, int[]> partitionsByMember = getPartitionAssignment(nodeEngine, memberInfos);
+        Map<Address, int[]> partitionsByAddress =
                 partitionsByMember.entrySet().stream().collect(toMap(en -> en.getKey().getAddress(), Entry::getValue));
-        final List<Address> addresses = toList(partitionsByMember.keySet(), MemberInfo::getAddress);
-        final int clusterSize = partitionsByMember.size();
+        int clusterSize = partitionsByMember.size();
         final boolean isJobDistributed = clusterSize > 1;
         final EdgeConfig defaultEdgeConfig = nodeEngine.getConfig().getJetConfig().getDefaultEdgeConfig();
         final Map<MemberInfo, ExecutionPlan> plans = new HashMap<>();
         int memberIndex = 0;
+        Set<Integer> requiredPartitions = jobConfig.getArgument(KEY_REQUIRED_PARTITIONS);
+        if (requiredPartitions != null) {
+            Map<MemberInfo, int[]> requiredMembers = new HashMap<>();
+            for (Entry<MemberInfo, int[]> entry : partitionsByMember.entrySet()) {
+                for (int partitionId : entry.getValue()) {
+                    if (requireNonNull(requiredPartitions).contains(partitionId)) {
+                        requiredMembers.putIfAbsent(entry.getKey(), entry.getValue());
+                    }
+                }
+            }
+            partitionsByMember = requiredMembers;
+            partitionsByAddress = requiredMembers.entrySet()
+                    .stream()
+                    .collect(toMap(en -> en.getKey().getAddress(), Entry::getValue));
+            clusterSize = partitionsByAddress.size();
+        }
```

Comment:
This should go into `getPartitionAssignment`, where it can be done more efficiently.

E.g. we can avoid computing the `partitionsByAddress` twice (line 88 and 106) when required partitions are non-null.

Also, Krzysztof S. benchmarked and did some micro-optimizations here, I would like to see this re-done with this new version.

Reply:
> Also, Krzysztof S. benchmarked and did some micro-optimizations here, I would like to see this re-done with this new version.

It will be done when whole solution will be assembled (both jet & sql part)

Reply:
Ok, I can accept that. Still, the code can be made more efficient right away. Let's do that please.

Reply:
@frant-hartm  sure, did it here : https://github.com/hazelcast/hazelcast/pull/24689/commits/7a2d295e40600cfd3d27c99acdb384b5fe396b98